lib/cchain/provider/provider.go

```diff
@@ -179,9 +183,18 @@ func (p Provider) stream(
 	srcChain := p.chainNamer(chainVer)
 	ctx := log.WithCtx(in, "src_chain", srcChain, "worker", workerName)
 
+	// cachedOffsetHeight is used to store the height at which the previous attestation offset was found
+	var cachedOffsetHeight uint64
+
 	deps := stream.Deps[xchain.Attestation]{
 		FetchBatch: func(ctx context.Context, _ uint64, offset uint64) ([]xchain.Attestation, error) {
-			return p.fetch(ctx, chainVer, offset)
+			atts, foundHeight, err := p.fetch(ctx, chainVer, offset, cachedOffsetHeight)
+			if err != nil {
+				return nil, err
+			}
+			cachedOffsetHeight = foundHeight
```

Comment:
There is a bug here...

Reply:
hint: you need to make the contracts of your functions clearer in comments, specifically edge cases, then callers need to handle those edge cases.